keps/1224-lending-limit/README.md

```diff
@@ -0,0 +1,162 @@
+
+## Motivation
+
+Sometimes we want to keep some resources for guaranteed usage, so that when new jobs come into queue, they can be admitted immediately.
+
+Under the current implementation, we are using `BorrowingLimit` to define the maximum amount of quota that this ClusterQueue is allowed to borrow. But this may cause another ClusterQueue in the same cohort to run out of resources.
+
+Even if we set the `Preemption`, it still needs some time and spends a lot of unnecessary cost.
+
+So we need a reservation design for resource requests and security reasons: `LendingLimit`, to claim the quota allowed to lend, reserve a certain amount of resources to ensure that they will never be borrowed.
+
+### Goals
+
+- Implement `LendingLimit`, users can have a reservation of guaranteed resource by claiming the `LendingLimit`.
+
+### Non-Goals
+
+- Replace `BorrowingLimit` to some extent in the future.
+
+## Proposal
+
+In this proposal, `LendingLimit` is defined. The `ClusterQueue` will be limited to lend the specified quota to other ClusterQueues in the same cohort.
+
+### User Stories (Optional)
+
+#### Story 1
+
+In order to ensure the full utilization of resources, we generally set `BorrowingLimit` to max, but this may cause a `ClusterQueue` to run out of its all resources, and make the new incoming job slow to response for the slow preemption. This could be worse in a competitive cluster, jobs will borrow resources and be reclaimed over and over.
+
+So we want to reserve some resources for a `ClusterQueue`, so that any incoming jobs in the  `ClusterQueue` can get admitted immediately.
+
+### Notes/Constraints/Caveats (Optional)
+
+With both BorrowingLimit and LendingLimit configured, one clusterQueue may not be able to borrow up to the limit just because we reserved the lending limit quota of resource.
+
+To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.
+
+### Risks and Mitigations
+
+None.
+
+## Design Details
+
+### Kueue LendingLimit API
+
+Modify ResourceQuota API object:
+
+```go
+type ResourceQuota struct {
+	[...]
+
+	// lendingLimit is the maximum amount of unused quota for the [flavor, resource]
+	// combination that this ClusterQueue can lend to other ClusterQueues in the same cohort.
+	// In total, at a given time, ClusterQueue reserves for its exclusive use
+	// a quantity of quota equals to nominalQuota - lendingLimit.
+	// If null, it means that there is no lending limit.
+	// If not null, it must be non-negative.
+	// lendingLimit must be null if spec.cohort is empty.
+	// +optional
+	LendingLimit *resource.Quantity `json:"lendingLimit,omitempty"`
+}
+```
+
+#### Note
+
+We have considered adding this status field, but deprecated it in the end. Because unused resources from multiple CQs compose a single pool of shareable resources, and we cannot precisely calculate this value.
```

Comment:
```suggestion
We have considered adding this status field, but discarded it. Because unused resources from multiple CQs compose a single pool of shareable resources, we cannot precisely calculate this value.
```